### PR TITLE
feat(User): Basic user preferences, GET User by attribute.

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
### Checklist

- [X] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [na] The description lists all applicable issues this PR seeks to resolve
- [na] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR adds the following:
* An api route to get entities by attribute, used in the new MOD UI to retrieve user data using an auth0 id. (The Mongo ID is not stored or communicated by Auth0).
* User fields and types related to creating user accounts on middleware.

### CODE REVIEW

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: "The attribute the entity to search." This doesn't make sense.

File: src/main/java/org/opentripplanner/middleware/models/User.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/models/User.java]: Does the User class now contain all required fields? If so, perhaps remove the TODO (on line 16).

File: src/main/java/org/opentripplanner/middleware/models/UserNotificationChannel.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/models/UserNotificationChannel.java]: If we decide that enums can't be saved to Mongo this will need to be changed.

File: src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java]: This is the only method that has this amount of detail in the comment (@param and @return). I'm not sure what the policy is on this, but probably easier to remove to be inline with the other comments?

**Review by br648 — COMMENTED** [review]
Some minor comments. All good otherwise.

**Comment by binh-dam-ibigroup** [comment]
Note: last commit will convert `notificationChannel` into `notificationChannels` to allow notifications with both email and sms.

File: src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java]: I would like this to be named `getFirstByFieldValue` to be clear that it is returning a single record and getting the first result whatever that may be.

File: src/main/java/org/opentripplanner/middleware/models/UserLocation.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/models/UserLocation.java:25]: Needs `lat`, `lon` added.

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Would like this renamed to `getFirstObjectByFieldValue` to indicate that it is getting the first record that matches a field value.

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: I'd like this renamed to `getFirstByFieldValue` to indicate that this is getting the first record that matches the given field value.

**Review by evansiroky — CHANGES_REQUESTED** [review]
See comments about renaming things to indicate that certain methods are returning the first item matching a field value. I feel this is important because the first result from a search on a field without distinct values could end up returning inconsistent records.

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Each of these get API endpoints don't have any security associated with them. Any requester could just send over an arbitrary Auth0 user id and it would return the user object for another user.

**Review by evansiroky — CHANGES_REQUESTED** [review]
Just discovered a big security risk, so I'm marking this PR as "requesting changes" until all get methods have authorization checks added to ensure data is returned only when proper authorization can be confirmed.

Also, it appears that the deletion methods need proper authorization checks as well.

**Comment by binh-dam-ibigroup** [comment]
Caution: On parent branch `add-sub-users`, a similar "filtered query" was introduced. Need to use it. See https://github.com/ibi-group/otp-middleware/blob/77b9e500c1f308690e41bec6cbd893567aaf3667/src/main/java/org/opentripplanner/middleware/auth/Auth0UserProfile.java#L43

**Comment by binh-dam-ibigroup** [comment]
> Just discovered a big security risk, so I'm marking this PR as "requesting changes" until all get methods have authorization checks added to ensure data is returned only when proper authorization can be confirmed.
> 
> Also, it appears that the deletion methods need proper authorization checks as well.

Suggestion: address the quoted comment in #13 or #14. Thoughts? @evansiroky and @landonreed

File: src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java]: I think this method will become `getOneFiltered` once #14 is merged.

File: src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java]: Method is redundant and I went ahead and removed it.

File: src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java]: Method is redundant. I went ahead and removed it.

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: This is too much abstraction to save just a little bit of repeated code. It makes things less readable, so I think the original way of doing this in `getObjectForId` is preferred.

Comment (evansiroky) [reply src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Or maybe a hybrid approach would be better where you can change this method to be something like:

```java
private void haltIfNull(T object, Request req, String message) {
    if (object == null) {
        logMessageAndHalt(req, HttpStatus.NOT_FOUND_404, message);
    }
}
```

**Review by evansiroky — CHANGES_REQUESTED** [review]
See https://github.com/ibi-group/otp-middleware/pull/15/files#r426110256 and https://github.com/ibi-group/otp-middleware/pull/15/files#r426108133 and https://github.com/ibi-group/otp-middleware/pull/15/files#r428252282.

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: @evansiroky This is addressed by #21 in https://github.com/ibi-group/otp-middleware/blob/b827e04b81d9b2fd31a7b558ab9f3b0cd846a473/src/main/java/org/opentripplanner/middleware/auth/Auth0Utils.java#L15 so we *will* be making those checks.

Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Actually, it is also handled by #13 in https://github.com/ibi-group/otp-middleware/blob/94dcefa65446530ba5854afb442aac7d416c5312/src/main/java/org/opentripplanner/middleware/auth/Auth0Connection.java#L38

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: `getOneByField` is undefined

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (evansiroky) [reply src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Ok, I'll overlook this for now since #21 has needed checks.

**Review by evansiroky — CHANGES_REQUESTED** [review]
Travis builds failing probably due to undefined method call. Probably good to go once that is fixed.

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Thanks for pointing it out!

**Review by evansiroky — DISMISSED** [review]

**Review by br648 — DISMISSED** [review]
Looks good!

**Review by br648 — DISMISSED** [review]
All good and unti test passes.

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Can we rename this to `getRequiredParamFromRequest` and add a note that this method will halt the request if the param is not provided?

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Please rename `EndPoint` to `Endpoint` to maintain consistency.

File: src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java]: Please rename `EndPoint` to `Endpoint` to maintain consistency.

File: src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:98]: This should return an HTTP 403 error since it seems that whatever request this is will fail if an authenticated user is not present in the request.

File: src/test/java/org/opentripplanner/middleware/auth/Auth0UsersTest.java
Comment (evansiroky) [inline src/test/java/org/opentripplanner/middleware/auth/Auth0UsersTest.java]: Since this is dependent on the time, I recommend putting in a Thead.sleep or something to be extra sure that some amount of time has elapsed.

**Review by evansiroky — CHANGES_REQUESTED** [review]
See most recent comments.

**Review by br648 — DISMISSED** [review]
Looks good.

File: src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:98]: @evansiroky Please describe a situation where this would be the case.

For users who just created an Auth0 login and have an `Auth0UserProfile`,
but have not completed the account setup form yet, the `OtpUser` profile would naturally be 404 not found until they complete the account setup form and you wouldn't want to 403 forbid that user to try to retrieve their account.

File: src/main/java/org/opentripplanner/middleware/models/OtpUser.java
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/models/OtpUser.java]: @binh-dam-ibigroup, is this field populated from the Auth0 field? We're not allowing users to create an account if their email has not already been verified, correct? This suggests to me that this field is no longer required.

File: src/main/java/org/opentripplanner/middleware/models/OtpUser.java
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/models/OtpUser.java:18]: Please add TODO because we have not yet added code to verify the phone number.

File: src/main/java/org/opentripplanner/middleware/models/OtpUser.java
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/models/OtpUser.java]: This comment is very cryptic. Please just bump the mongo version if that resolves the enum storage issue.

File: src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java
Comment (evansiroky) [reply src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:98]: Huh, I guess this one could go either way.

**Review by evansiroky — DISMISSED** [review]

**Review by landonreed — CHANGES_REQUESTED** [review]
Looks good, just a few changes.

File: src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java]: I think Evan mentioned this, lowercase 'point' so baseEndpoint.

**Review by br648 — DISMISSED** [review]
One very minor item, apart from that all good!

File: src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java]: Fixed. Thanks.

File: src/main/java/org/opentripplanner/middleware/models/OtpUser.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/OtpUser.java]: I will remove. We don't use this field and get that information straight from Auth0.

Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/OtpUser.java]: Same for recent locations.

File: src/main/java/org/opentripplanner/middleware/models/OtpUser.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/OtpUser.java]: Updating the comment. The Mongo driver supports the desired functionality, however there is additional setup needed.

**Review by br648 — APPROVED** [review]
Looks good!

**Review by evansiroky — APPROVED** [review]

**Review by landonreed — APPROVED** [review]
Good stuff. Thanks for the changes.

**Pull request merged**
